pkg/controllers/raycluster_webhook_test.go

```diff
@@ -0,0 +1,393 @@
+			}
+		}
+
+		// Check each worker group independently
+		for _, workerSpec := range validRayCluster.Spec.WorkerGroupSpecs {
+			workerInitContainers := workerSpec.Template.Spec.InitContainers
+			checkInitContainers(workerInitContainers)
+		}
+	})
+}
+
+func TestValidateCreate(t *testing.T) {
+	test := NewTest(t)
+
+	validRayCluster := &rayv1.RayCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-valid-raycluster",
+			Namespace: namespace,
+		},
+		Spec: rayv1.RayClusterSpec{
+			HeadGroupSpec: rayv1.HeadGroupSpec{
+				Template: corev1.PodTemplateSpec{
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  oauthProxyContainerName,
+								Image: "registry.redhat.io/openshift4/ose-oauth-proxy@sha256:1ea6a01bf3e63cdcf125c6064cbd4a4a270deaf0f157b3eabb78f60556840366",
+								Ports: []corev1.ContainerPort{
+									{ContainerPort: 8443, Name: "oauth-proxy"},
+								},
+								Env: []corev1.EnvVar{
+									{
+										Name: "COOKIE_SECRET",
+										ValueFrom: &corev1.EnvVarSource{
+											SecretKeyRef: &corev1.SecretKeySelector{
+												LocalObjectReference: corev1.LocalObjectReference{
+													Name: "test-valid-raycluster" + "-oauth-config",
+												},
+												Key: "cookie_secret",
+											},
+										},
+									},
+								},
+								Args: []string{
+									"--https-address=:8443",
+									"--provider=openshift",
+									"--openshift-service-account=" + "test-valid-raycluster" + "-oauth-proxy",
+									"--upstream=http://localhost:8265",
+									"--tls-cert=/etc/tls/private/tls.crt",
+									"--tls-key=/etc/tls/private/tls.key",
+									"--cookie-secret=$(COOKIE_SECRET)",
+									"--openshift-delegate-urls={\"/\":{\"resource\":\"pods\",\"namespace\":\"" + namespace + "\",\"verb\":\"get\"}}",
+								},
+								VolumeMounts: []corev1.VolumeMount{
+									{
+										Name:      oauthProxyVolumeName,
+										MountPath: "/etc/tls/private",
+										ReadOnly:  true,
+									},
+								},
+							},
+						},
+						Volumes: []corev1.Volume{
+							{
+								Name: oauthProxyVolumeName,
+								VolumeSource: corev1.VolumeSource{
+									Secret: &corev1.SecretVolumeSource{
+										SecretName: "test-valid-raycluster" + "-proxy-tls-secret",
+									},
+								},
+							},
+						},
+						ServiceAccountName: "test-valid-raycluster" + "-oauth-proxy",
+					},
+				},
+				RayStartParams: map[string]string{},
+			},
+		},
+	}
+
+	// Create the RayClusters
+	test.Client().Ray().RayV1().RayClusters(namespace).Create(test.Ctx(), validRayCluster, metav1.CreateOptions{})
+
+	// Call to ValidateCreate function is made
+	warnings, err := rcWebhook.ValidateCreate(test.Ctx(), runtime.Object(validRayCluster))
+	t.Run("Expected no warnings or errors on call to ValidateCreate function", func(t *testing.T) {
+		g := gomega.NewWithT(t)
+		g.Expect(warnings).To(gomega.BeNil())
+		g.Expect(err).To(gomega.BeNil())
+	})
```

Comment:
@sutaakar I only then realised that this basically covers all test cases. Perhaps the tests after this one are not needed?

Also, If I do the same for the Default function, would all the other tests/assertions not be required?

Open to thoughts on this before continuing, thanks!

Reply:
Thinking perhaps the assertions made for the Default (mutating) function is required to compare outputs. But, for the validating functions such as `ValidateCreate`, as long as no errors are returned (meaning all elements exist in the defined RayCluster), the validation should succeed.

Reply:
For `ValidateCreate` it is enough to have one positive test, providing negative tests to validate incorrect modification and its detection.

Reply:
Or possibly to have several tests, providing valid RayClusters, i.e. empty workerGroupSpecs and such.

Reply:
Hopefully, latest changes in this PR reflects the above. - One positive test case, providing negative tests for manipulated fields. Thanks!